---
title: "Basic Deconvolution"
author: "Gregory Hunt"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Basic Deconvolution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

We can show a basic deconvolution example by working with a data set created from rats. This is available on GEO with accession [GSE19830](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19830).

First let's download the data set and read it in using the ``affy`` package.
```{r}
rat.url <- "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE19nnn/GSE19830/suppl/GSE19830_RAW.tar"
rat.dir <- "rat_data"
dir.create(rat.dir, showWarnings = FALSE)
rat.file <- "rat.tar"
```

```{r,warning=FALSE,cache=FALSE,eval=TRUE}
download.file(rat.url, destfile = rat.file)
untar(rat.file, exdir = rat.dir)
```

```{r}
suppressPackageStartupMessages(library('affy'))
rat <- ReadAffy(celfile.path = rat.dir)
```

```{r,echo=FALSE,results='hide',cache=FALSE,eval=TRUE}
file.remove(rat.file)
file.remove(paste(rat.dir,dir(rat.dir),sep="/"))
unlink(rat.dir,force=TRUE,recursive=TRUE)
```

This is a data set of rat brain, liver and lung cells that have been mixed together in various proportions and then analyzed with by DNA microarrays. The true mixing proportions are encoded in the following matrix. 

```{r,cache=FALSE}
rat.truth = matrix(c(0,0,0,1,1,1,0,0,0,0.25,0.25,0.25,0.05,0.05,0.05,0.7,0.7,0.7,0.25,0.25,0.25,0.45,0.45,0.45,0.2,0.2,0.2,0.3,0.3,0.3,0.3,0.3,0.3,0.4,0.4,0.4,0.35,0.35,0.35,0.34,0.34,0.34,1,1,1,0,0,0,0,0,0,0.05,0.05,0.05,0.7,0.7,0.7,0.25,0.25,0.25,0.7,0.7,0.7,0.45,0.45,0.45,0.55,0.55,0.55,0.5,0.5,0.5,0.55,0.55,0.55,0.5,0.5,0.5,0.6,0.6,0.6,0.65,0.65,0.65,0,0,0,0,0,0,1,1,1,0.7,0.7,0.7,0.25,0.25,0.25,0.05,0.05,0.05,0.05,0.05,0.05,0.1,0.1,0.1,0.25,0.25,0.25,0.2,0.2,0.2,0.15,0.15,0.15,0.1,0.1,0.1,0.05,0.05,0.05,0.01,0.01,0.01),ncol=3,byrow=FALSE)
head(rat.truth)
```

Each row is a sample and each column gives the mixing proportions of the cell type. From this we can extract out the pure samples of each of the three cell types. 
```{r,cache=FALSE}
pure_samples <- lapply(1:3, function(i) {
    which(rat.truth[, i] == 1)
})
pure_samples
```

Now the deconvolution algorithm can work with many different types of data. Ultimately we just need a matrix of expressions for each experiment. Let's first look at the probe level data. 

```{r}
Y <- log2(t(affy::intensity(rat)))
Y[1:4,1:4]
```
Each row is a different experiment and each column is the log expression measurement of a particular oligonucleotide. 

Here is an example of how to run the deconvolution method. First we ask to find the marker genes. We pass in the data matrix, ``Y``, the list of pure samples for each type, ``pure_samples``, the data type, ``data_type``, and the method used to rank markers, ``method``, to the function ``find_markers``. The default ``method`` is "eta".
```{r}
library('deconv')
marker_list = find_markers(Y,pure_samples,data_type="microarray-probe",method='eta')
```
The function ``find_markers`` returns a list of two elements. The first is ``L`` which is a list of the same length as ``pure_samples`` with the $i^{th}$ element being markers for the $i^{th}$ type ranked in decreasing order of importance (best markers listed first). The second element of the list ``find_markers`` returns is a list of the same structure named ``V``. This list contains the computed value for each marker that was used to rank the markers. The meaning of this value depends on the choice of ``method``. For ``method=eta`` the values in ``V`` are the negative of the ratio of the estimated amount the particular marker is expressed in all other types to the amount that the marker is expressed in the $i^{th}$ type. 

After we have ranked our markers we need to determine how many markers to use for each cell type. One way to do this is to threshold the values in ``V`` such that we only use a marker if the computed value is above a certain threshold. For example, we can set our threshold at ``.005`` so that we only keep markers if they are expressed in the cell type they mark at an estimated 200 times higher than how much they are expressed in all other cell types combined. 
```{r}
thresh = .005
K = length(pure_samples)
n_choose = sapply(1:K,function(i){max(which(marker_list$V[[i]] > (-thresh)))})
```
for this example the number we choose to keep for each cell type is assigned to ``n_choose``,
```{r}
n_choose
```

Now that we have ranked our markers for each type and chosen how many to use for each type we can run the deconvolution algorithm.
```{r}
marks = marker_list$L
dc <- deconv(Y,pure_samples,n_choose,data_type='microarray-probe',full_markers=marks)
```
providing to the ``deconv`` function ``Y``, our microarray data, the list of ``pure_samples``, the ``data_type``, the number of top markers to use for each cell type, ``n_choose``, and the list of ranked markers for each type to the ``full_markers`` argument. 

The ``deconv`` function returns to us a list with elements ``estimates``, the estimated mixing proportions for each type for each sample, ``markers``, the markers used for each cell type, ``n_choose``, how many markers we used for each cell type, and ``gamma``, the value of the sensitivity parameter used. 

We can plot our estimates against the known truth as follows
```{r,results='asis',fig.height=5,fig.width=5}
phats <- dc$estimates
plot(rat.truth,phats,xlab="Truth",ylab="Estimates",xlim=c(0,1),ylim=c(0,1))
abline(coef=c(0,1))
```

If we want we can specify the value of the sensivity parameter ``gamma`` specificaly instead of choosing it based upon the ``data_type``. For example, 
```{r,results='asis',fig.height=5,fig.width=5}
dc <- deconv(Y,pure_samples,n_choose,gamma=.7,full_markers=marks)
phats <- dc$estimates
plot(rat.truth,phats,xlab="Truth",ylab="Estimates",xlim=c(0,1),ylim=c(0,1))
abline(coef=c(0,1))
```

We can view the pre-selected values for ``gamma`` from the function ``get_gamma``
```{r}
get_gamma('microarray-probe')
get_gamma('microarray-gene')
get_gamma('rna-seq')
```
We can also specify the number of markers to be the same for each cell type by providing a single number to ``n_choose``
```{r,results='asis',fig.height=5,fig.width=5}
dc <- deconv(Y,pure_samples,200,data_type='microarray-probe',full_markers=marks)
phats <- dc$estimates
plot(rat.truth,phats,xlab="Truth",ylab="Estimates",xlim=c(0,1),ylim=c(0,1))
abline(coef=c(0,1))
```

We can test different methods of choosing markers by specifying the ``marker_method`` argument.
```{r,results='asis',fig.height=5,fig.width=5}
marker_list = find_markers(Y,pure_samples,data_type="microarray-probe",method='eta')
marks = marker_list$L
thresh = .005
K = length(pure_samples)
n_choose = sapply(1:K,function(i){max(which(marker_list$V[[i]] > (-thresh)))})

dc <- deconv(Y, pure_samples,n_choose,data_type='microarray-probe',marker_method = 'eta')
phats <- dc$estimate
plot(rat.truth,phats,xlab="Truth",ylab="Estimates",xlim=c(0,1),ylim=c(0,1))
abline(coef=c(0,1))

dc2 <- deconv(Y, pure_samples,n_choose,data_type='microarray-probe',marker_method = 'diff')
phats2 <- dc2$estimates
points(rat.truth,phats2,col='blue')

dc3 <- deconv(Y, pure_samples,n_choose,data_type='microarray-probe',marker_method = 'regression')
phats3 <- dc3$estimates
points(rat.truth,phats3,col='red')
```

We can also run the algorithm on RMA summarized expressions. 
```{r,warning=FALSE,results='hide',cache=FALSE}
Y <- data.frame(affy::rma(rat,verbose=FALSE))

marker_list = find_markers(Y,pure_samples,data_type="microarray-gene",method='eta')
marks = marker_list$L
thresh = .005
K = length(pure_samples)
n_choose = sapply(1:K,function(i){max(which(marker_list$V[[i]] > (-thresh)))})

dc <- deconv(Y, pure_samples,n_choose,data_type='microarray-probe',marker_method = 'eta')
phats <- dc$estimate
plot(rat.truth,phats,xlab="Truth",ylab="Estimates",xlim=c(0,1),ylim=c(0,1))
abline(coef=c(0,1))
```
